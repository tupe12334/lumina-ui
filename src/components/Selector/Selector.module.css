/* Base selector container */
.selector {
  position: relative;
  display: inline-block;
  min-width: 200px;
}

/* Trigger button */
.trigger {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  background-color: #ffffff;
  border: 1px solid #d1d5db;
  border-radius: 0.375rem;
  padding: 0.5rem 0.75rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #374151;
  cursor: pointer;
  transition: all 0.15s ease-in-out;
  text-align: left;
}

.trigger:focus {
  outline: none;
  border-color: #3b82f6;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.1);
}

.trigger:hover:not(:disabled) {
  border-color: #9ca3af;
}

.trigger:disabled {
  background-color: #f9fafb;
  color: #9ca3af;
  cursor: not-allowed;
}

/* Size variants */
.selector--sm .trigger {
  padding: 0.25rem 0.5rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

.selector--md .trigger {
  padding: 0.5rem 0.75rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.selector--lg .trigger {
  padding: 0.75rem 1rem;
  font-size: 1rem;
  line-height: 1.5rem;
}

/* Variant styles */
.selector--error .trigger {
  border-color: #ef4444;
}

.selector--error .trigger:focus {
  border-color: #ef4444;
  box-shadow: 0 0 0 3px rgba(239, 68, 68, 0.1);
}

.selector--success .trigger {
  border-color: #10b981;
}

.selector--success .trigger:focus {
  border-color: #10b981;
  box-shadow: 0 0 0 3px rgba(16, 185, 129, 0.1);
}

/* Value and placeholder */
.value {
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.placeholder {
  color: #9ca3af;
}

/* Dropdown icon */
.icon {
  margin-left: 0.5rem;
  font-size: 0.75rem;
  color: #6b7280;
  transition: transform 0.15s ease-in-out;
  user-select: none;
}

.iconOpen {
  transform: rotate(180deg);
}

/* Listbox */
.listbox {
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 50;
  margin-top: 0.25rem;
  background-color: #ffffff;
  border: 1px solid #d1d5db;
  border-radius: 0.375rem;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);
  overflow-y: auto;
  padding: 0.25rem 0;
  list-style: none;
  margin: 0;
}

/* Options */
.option {
  display: flex;
  flex-direction: column;
  padding: 0.5rem 0.75rem;
  cursor: pointer;
  color: #374151;
  transition: background-color 0.15s ease-in-out;
}

.option:hover,
.optionFocused {
  background-color: #f3f4f6;
}

.optionSelected {
  background-color: #3b82f6;
  color: #ffffff;
}

.optionSelected:hover,
.optionSelected.optionFocused {
  background-color: #2563eb;
}

.optionDisabled {
  color: #9ca3af;
  cursor: not-allowed;
  background-color: transparent !important;
}

.optionLabel {
  font-weight: 500;
}

.optionDescription {
  font-size: 0.75rem;
  color: #6b7280;
  margin-top: 0.125rem;
}

.optionSelected .optionDescription {
  color: #e5e7eb;
}

/* No options state */
.noOptions {
  padding: 0.75rem;
  text-align: center;
  color: #9ca3af;
  font-size: 0.875rem;
  font-style: italic;
}

/* Open state */
.selector--open .trigger {
  border-color: #3b82f6;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.1);
}

/* Disabled state */
.selector--disabled .trigger {
  background-color: #f9fafb;
  color: #9ca3af;
  cursor: not-allowed;
}

.selector--disabled .icon {
  color: #9ca3af;
}