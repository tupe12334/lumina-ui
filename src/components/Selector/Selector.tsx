import React, { useState, useRef, useEffect } from 'react'
import { clsx } from 'clsx'
import styles from './Selector.module.css'

export interface SelectorOption {
  value: string
  label: string
  disabled?: boolean
}

export interface SimpleSelectorProps {
  options: SelectorOption[]
  value?: string
  placeholder?: string
  disabled?: boolean
  className?: string
  onChange?: (value: string) => void
}

function useSelectorState(options: SelectorOption[], value?: string, onChange?: (value: string) => void) {
  const [isOpen, setIsOpen] = useState(false)
  const [selectedValue, setSelectedValue] = useState(value)
  const selectedOption = options.find(option => option.value === selectedValue)
  const displayText = selectedOption ? selectedOption.label : 'Select option'
  const handleSelect = (optionValue: string) => {
    setSelectedValue(optionValue)
    setIsOpen(false)
    if (onChange) onChange(optionValue)
  }
  useEffect(() => setSelectedValue(value), [value])
  return { isOpen, setIsOpen, selectedValue, displayText, handleSelect }
}

export const Selector = React.forwardRef<HTMLDivElement, SimpleSelectorProps>(
  (props, ref) => {
    const { options, value, placeholder, disabled, className, onChange } = props
    const buttonRef = useRef<HTMLButtonElement>(null)
    const listRef = useRef<HTMLUListElement>(null)

    const { isOpen, setIsOpen, selectedValue, displayText, handleSelect } = useSelectorState(options, value, onChange)
    const finalDisplayText = displayText !== 'Select option' ? displayText : placeholder || 'Select option'
    const handleToggle = () => { if (!disabled) setIsOpen(!isOpen) }
    const handleKeyDown = (event: React.KeyboardEvent) => { if (event.key === 'Escape') setIsOpen(false) }

    useEffect(() => {
      function handleClickOutside(event: MouseEvent) {
        const target = event.target
        if (
          isOpen &&
          buttonRef.current &&
          listRef.current &&
          target instanceof Node &&
          !buttonRef.current.contains(target) &&
          !listRef.current.contains(target)
        ) {
          setIsOpen(false)
        }
      }

      document.addEventListener('mousedown', handleClickOutside)
      return () => document.removeEventListener('mousedown', handleClickOutside)
    }, [isOpen, setIsOpen])

    return (
      <div ref={ref} className={clsx(styles.selector, className)}>
        <button
          ref={buttonRef}
          type="button"
          disabled={disabled}
          className={clsx(styles.trigger)}
          onClick={handleToggle}
          onKeyDown={handleKeyDown}
        >
          <span className={clsx(styles.value)}>{finalDisplayText}</span>
          <span className={clsx(styles.icon)} aria-hidden="true">â–¼</span>
        </button>

        {isOpen && (
          <ul ref={listRef} className={clsx(styles.listbox)}>
            {options.map((option) => (
              <li
                key={option.value}
                className={clsx(styles.option, {
                  [styles.optionSelected]: option.value === selectedValue,
                  [styles.optionDisabled]: option.disabled
                })}
                onClick={() => !option.disabled && handleSelect(option.value)}
              >
                <span className={clsx(styles.optionLabel)}>{option.label}</span>
              </li>
            ))}
          </ul>
        )}
      </div>
    )
  }
)

Selector.displayName = 'Selector'