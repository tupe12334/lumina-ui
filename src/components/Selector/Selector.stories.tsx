import type { Meta, StoryObj } from '@storybook/react'
import { Selector } from './Selector'

const meta: Meta<typeof Selector> = {
  title: 'Components/Selector',
  component: Selector,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    size: {
      control: 'select',
      options: ['sm', 'md', 'lg'],
    },
    variant: {
      control: 'select',
      options: ['default', 'error', 'success'],
    },
    disabled: {
      control: 'boolean',
    },
    error: {
      control: 'boolean',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

const sampleOptions = [
  { value: 'apple', label: 'Apple', description: 'A sweet red fruit' },
  { value: 'banana', label: 'Banana', description: 'A yellow tropical fruit' },
  { value: 'orange', label: 'Orange', description: 'A citrus fruit' },
  { value: 'grape', label: 'Grape', description: 'Small purple fruits' },
  { value: 'strawberry', label: 'Strawberry', description: 'A red berry' },
  { value: 'disabled', label: 'Disabled Option', description: 'This option is disabled', disabled: true },
]

export const Default: Story = {
  args: {
    options: sampleOptions,
    placeholder: 'Select a fruit',
  },
}

export const WithValue: Story = {
  args: {
    options: sampleOptions,
    value: 'apple',
    placeholder: 'Select a fruit',
  },
}

export const Small: Story = {
  args: {
    options: sampleOptions,
    size: 'sm',
    placeholder: 'Select a fruit',
  },
}

export const Large: Story = {
  args: {
    options: sampleOptions,
    size: 'lg',
    placeholder: 'Select a fruit',
  },
}

export const ErrorState: Story = {
  args: {
    options: sampleOptions,
    variant: 'error',
    placeholder: 'Select a fruit',
  },
}

export const SuccessState: Story = {
  args: {
    options: sampleOptions,
    variant: 'success',
    value: 'apple',
    placeholder: 'Select a fruit',
  },
}

export const Disabled: Story = {
  args: {
    options: sampleOptions,
    disabled: true,
    placeholder: 'Select a fruit',
  },
}

export const WithSearchAndClear: Story = {
  args: {
    options: sampleOptions,
    placeholder: 'Search and select a fruit',
    config: {
      searchable: true,
      clearable: true,
    },
  },
}

export const ManyOptions: Story = {
  args: {
    options: Array.from({ length: 50 }, (_, i) => ({
      value: `option-${i}`,
      label: `Option ${i + 1}`,
      description: `Description for option ${i + 1}`,
    })),
    placeholder: 'Select from many options',
    config: {
      searchable: true,
      maxHeight: 200,
    },
  },
}