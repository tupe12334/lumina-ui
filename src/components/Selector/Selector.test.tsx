import { render, screen, fireEvent } from '@testing-library/react'
import { test, expect, vi } from 'vitest'
import { Selector } from './Selector'

const mockOptions = [
  { value: 'option1', label: 'Option 1' },
  { value: 'option2', label: 'Option 2' },
  { value: 'option3', label: 'Option 3', disabled: true },
]

test('Selector renders correctly with default props', () => {
  const { container } = render(<Selector options={mockOptions} />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly with selected value', () => {
  const { container } = render(<Selector options={mockOptions} value="option1" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly with placeholder', () => {
  const { container } = render(
    <Selector options={mockOptions} placeholder="Choose an option" />
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly with small size', () => {
  const { container } = render(<Selector options={mockOptions} size="sm" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly with large size', () => {
  const { container } = render(<Selector options={mockOptions} size="lg" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly with error variant', () => {
  const { container } = render(<Selector options={mockOptions} variant="error" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly with success variant', () => {
  const { container } = render(<Selector options={mockOptions} variant="success" />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector renders correctly when disabled', () => {
  const { container } = render(<Selector options={mockOptions} disabled />)
  expect(container.firstChild).toMatchSnapshot()
})

test('Selector calls onChange when option is selected', () => {
  const handleChange = vi.fn()
  render(<Selector options={mockOptions} onChange={handleChange} />)

  const trigger = screen.getByRole('button')
  fireEvent.click(trigger)

  const option = screen.getByText('Option 1')
  fireEvent.click(option)

  expect(handleChange).toHaveBeenCalledWith('option1')
})

test('Selector opens dropdown when trigger is clicked', () => {
  render(<Selector options={mockOptions} />)

  const trigger = screen.getByRole('button')
  fireEvent.click(trigger)

  expect(screen.getByRole('listbox')).toBeInTheDocument()
})

test('Selector shows placeholder when no value selected', () => {
  render(<Selector options={mockOptions} placeholder="Select option" />)

  expect(screen.getByText('Select option')).toBeInTheDocument()
})

test('Selector shows selected option label', () => {
  render(<Selector options={mockOptions} value="option1" />)

  expect(screen.getByText('Option 1')).toBeInTheDocument()
})

test('Selector renders correctly with custom className', () => {
  const { container } = render(
    <Selector options={mockOptions} className="custom-class" />
  )
  expect(container.firstChild).toMatchSnapshot()
})