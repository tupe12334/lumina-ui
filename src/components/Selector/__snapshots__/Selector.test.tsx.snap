// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Selector renders correctly when disabled 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    disabled=""
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with custom className 1`] = `
<div
  class="_selector_729a27 custom-class"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with default props 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with error variant 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with large size 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with placeholder 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Choose an option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with selected value 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Option 1
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with small size 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;

exports[`Selector renders correctly with success variant 1`] = `
<div
  class="_selector_729a27"
>
  <button
    class="_trigger_729a27"
    type="button"
  >
    <span
      class="_value_729a27"
    >
      Select option
    </span>
    <span
      aria-hidden="true"
      class="_icon_729a27"
    >
      ▼
    </span>
  </button>
</div>
`;
